README.rst
===========
GraphTiler
===========

Display multiple configurable graphs at once with real-time data updates from a database.

* Graphs can be rearranged and resized.
* Multiple datasets can be displayed on the same graph with live updates.
* Layouts can be saved and restored.
* React & Redux frontend, SASS, Chart.js graphs.
* Quart (Flask compatible) backend with RESTful API and Cerberus for data validation.
* MongoDB database.
* For Python 3.7 and up
* MIT License


Demo
----

`Demonstration Video <https://youtu.be/7F43OCp4Ffk>`_


Installation
------------
1. Install `MongoDB <https://www.mongodb.com>`_ .

2. Clone this repository.

3. Run::

    python setup.py

4. Modify cfg.py to set your choice of IP address for the server and database.

5. From the install directory run the following replacing "YOUR_SERVER_ADDRESS" with the ip:port that you chose in step 4::

    hypercorn -b YOUR_SERVER_ADDRESS --reload --workers 1 GraphTiler:app


API Reference
-------------
[GET]  /category/

Returns a list of all categories.

.. code-block:: python

    from requests import get, post

    resp = get(url="http://localhost:7123/category/")
    print(resp.json())

[GET, POST]  /category/<name>

GET to get data for a specific category or POST to create a new category.

.. code-block:: python

    from requests import get, post

    cat_name = "fanSpeed"
    data = {
        "displayName": "Fan Speed",
        "units": "Rotations Per Minute",
        "abrvUnit": "RPM",
        "decimalPlaces": 0,
        "min": 0,
        "max": 10000,
    }

    # Create a new category.
    resp = post(url=f"http://localhost:7123/category/{cat_name}", json=data)
    print(resp.json())

    # Retrieve information about the newly created category.
    resp = get(url=f"http://localhost:7123/category/{cat_name}")
    print(resp.json())


[POST]  /modify-category/<name> :

Modify category information.

.. code-block:: python

    from requests import get, post

    cat_name = "fanSpeed"
    data = {
        "displayName": "Fan Rotation Speed",
        "units": "RPM",
        "abrvUnit": "RPM",
        "decimalPlaces": 2,
        "min": 10,
        "max": 20000,
    }

    resp = post( url=f"http://localhost:7123/modify-category/{cat_name}",
                 json=data )
    print(resp.json())


[POST]  /remove-category/<name>

Remove a category and all associated data points.

.. code-block:: python

    from requests import get, post

    cat_name = "fanSpeed"
    resp = post(url=f"http://localhost:7123/remove-category/{cat_name}")
    print(resp.json())


[POST]  /category/<name>/now :

Add a single data point for the current time.

.. code-block:: python

    from requests import get, post

    cat_name = "fanSpeed"
    data = {"reading": 500}

    # Get data for specific time points.
    # data = { "times": ["2019-10-22 09:15:00 -06:00",
    #                    "2019-10-22 09:16:05 -06:00"] }

    resp = post(url=f"http://localhost:7123/category/{cat_name}/now")
    print(resp.json())


[POST]  /category/<name>/add :

Add data points for the specified times.

.. code-block:: python

    from requests import get, post

    cat_name = "fanSpeed"
    data = {
        "readings": [
            {"time": "2019-10-22 09:15:00 -0600", "reading": 500},
            {"time": "2019-10-22 09:16:10 -0600", "reading": 600},
            {"time": "2019-10-22 09:17:20 -0600", "reading": 300},
            {"time": "2019-10-22 09:18:30 -0600", "reading": 400},
            {"time": "2019-10-22 09:14:40 -0600", "reading": 475},
            {"time": "2019-10-22 09:19:50 -0600", "reading": 800},
        ]
    }

    resp = post(url=f"http://localhost:7123/category/{cat_name}/add", json=data)
    print(resp.json())

[POST]  /category/<name>/get-points :

Get data points for a category for a time range or specific times.

.. code-block:: python

    from requests import get, post

    cat_name = "fanSpeed"
    data = {
        "range": {
            "start": "2019-10-22 09:00:00 -06:00",
            "end": "2019-10-22 14:00:00 -0600",
        }
    }

    # Get data for specific time points.
    # data = { "times": ["2019-10-22 09:15:00 -06:00",
    #                    "2019-10-22 09:16:05 -06:00"] }

    resp = post( url=f"http://localhost:7123/category/{cat_name}/get-points",
                 json=data )
    print(resp.json())


[POST]  /category/<name>/remove :

Remove data points for a category for a time range or specific times.

.. code-block:: python

    from requests import get, post

    cat_name = "fanSpeed"
    data = {
        "range": {
            "start": "2019-10-22 09:00:00 -06:00",
            "end": "2019-10-22 14:00:00 -0600",
        }
    }

    # Remove data points for specific time points.
    # data = { "times": ["2019-10-22 09:15:00 -06:00",
    #                    "2019-10-22 09:16:05 -06:00"] }

    resp = post( url=f"http://localhost:7123/category/{cat_name}/remove",
                 json=data )
    print(resp.json())


[POST]  /category/<name>/remove-all :

Remove all data points for a category.

.. code-block:: python

    from requests import get, post

    cat_name = "fanSpeed"

    resp = post(url=f"http://localhost:7123/category/{cat_name}/remove-all")
    print(resp.json())


CHANGELOG.rst
=========
Changelog
=========

0.9.0.0
-------
* Initial release.


lib/db_interface.py
from pymongo import MongoClient, UpdateOne, DESCENDING
from pymongo.errors import CollectionInvalid, ServerSelectionTimeoutError
from lib.funcs import no_errors
from lib.validation import ADD_SCHEMA
from lib.validation_funcs import str_to_datetime
from cerberus import Validator
from datetime import timezone, timedelta, datetime
from collections import defaultdict
from cfg import (
    MONGO_PORT,
    MONGO_ADDRESS,
    TIME_FORMAT,
    TIME_FORMAT_NO_TZ,
    TIME_MULTS,
)

KEY_ERR_MSG = "'time' and 'reading' are required keys for data points."
DATE_ERR_MSG = "Invalid date.  Date format must be "
"'YYYY-MM-DD HH:MM ±HHMM' where '±HHMM' is the time zone offset (ex. "
"'2020-01-20 18:45 -0600'). All numbers must have leading zeros and "
"hours are in 24 hour format."
READING_ERR_MSG = "Reading is not a number."
MIN_VAL_ERR_MSG = "Reading value is lower than the minimum allowed for this "
"category."
MAX_VAL_ERR_MSG = "Reading value is higher than the maximum allowed for this "
"category."


class DBInterface:
    """Interface for MongoDB."""

    def __init__(self):
        self._client = MongoClient(f"mongodb://{MONGO_ADDRESS}:{MONGO_PORT}")
        self._db = self._client.graphTiler
        self._add_vali = Validator(ADD_SCHEMA, require_all=True)

    def add_points(self, catname, data):
        """Add data points to the database."""
        errors, db_updates, added_points = [], [], []
        cat_info = self.get_category(catname)

        if not cat_info:
            errors.append(
                {"index": -1, "error": f'Category "{catname}" not found.'}
            )
            return {"errors": errors}
        min_val = cat_info["min"] if "min" in cat_info else None
        max_val = cat_info["max"] if "max" in cat_info else None

        if not self._add_vali.validate(data):
            errors.append({"index": -1, "error": self._add_vali.errors})
            return {"errors": errors}

        for i, point in enumerate(data["readings"]):
            if "time" not in point or "reading" not in point:
                errors.append({"index": i, "error": KEY_ERR_MSG})
                continue
            tm, reading = str_to_datetime(point["time"]), point["reading"]
            if tm is False:
                errors.append({"index": i, "error": DATE_ERR_MSG})
                continue
            point["time"] = tm
            if not isinstance(reading, int) and not isinstance(reading, float):
                errors.append({"index": i, "error": READING_ERR_MSG})
                continue
            if min_val is not None and reading < min_val:
                errors.append({"index": i, "error": MIN_VAL_ERR_MSG})
                continue
            elif max_val is not None and reading > max_val:
                errors.append({"index": i, "error": MAX_VAL_ERR_MSG})
                continue

            update = UpdateOne({"time": tm}, {"$set": point}, upsert=True)
            db_updates.append(update)
            added_points.append(point)

        if db_updates:
            try:
                self._db[f"catdata_default_{catname}"].bulk_write(db_updates)
            except ServerSelectionTimeoutError:
                return {"errors": {"server": "Could not connect to database"}}

        errors = None if not errors else errors
        return {"errors": errors, "added_points": added_points}

    def get_all_categories(self):
        """Get information on all current categories."""
        try:
            return self._db.categories.find({}, {"_id": 0})
        except ServerSelectionTimeoutError:
            return {"errors": {"server": "Could not connect to database"}}

    def get_category(self, catname):
        """Get information on a specific category."""
        try:
            return self._db.categories.find_one({"name": catname}, {"_id": 0})
        except ServerSelectionTimeoutError:
            return {"errors": {"server": "Could not connect to database"}}

    def add_category(self, catname, data):
        """Add categories to the database."""
        colname = f"catdata_default_{catname}"
        try:
            self._db.create_collection(colname)
        except CollectionInvalid:
            return {
                "errors": {
                    "name": f'A category named "{catname}" already exists.'
                }
            }
        except ServerSelectionTimeoutError:
            return {"errors": {"server": "Could not connect to database"}}

        data["name"] = catname
        try:
            self._db.categories.insert_one(data)
            self._db[colname].create_index([("time", DESCENDING)], unique=True)
            data.pop("_id", None)
            return no_errors()
        except ServerSelectionTimeoutError:
            return {"errors": {"server": "Could not connect to database"}}

    def modify_category(self, catname, data):
        """
        Modify an existing category.

        Does not make any changes to the data points for the category.
        """
        if not self.get_category(catname):
            return {
                "errors": {"name": f'Category "{catname}" does not exist.'}
            }
        try:
            self._db.categories.delete_one({"name": catname})
            data["name"] = catname
            self._db.categories.insert_one(data)
        except ServerSelectionTimeoutError:
            return {"errors": {"server": "Could not connect to database"}}
        return no_errors()

    def remove_category(self, catname):
        """Remove a category."""
        if self.get_category(catname) is None:
            return {"errors": f'Category "{catname}" does not exist.'}
        try:
            self._db.categories.delete_one({"name": catname})
            self._db[f"catdata_default_{catname}"].drop()
        except ServerSelectionTimeoutError:
            return {"errors": {"server": "Could not connect to database"}}
        return no_errors()

    def get_layout(self, name):
        """Get a saved layout."""
        return self._db.layouts.find_one({"name": name}, {"_id": 0})

    def get_all_layouts(self):
        """Get a list of all saved layouts."""
        filter = {"name": 1, "_id": 0}
        try:
            output = [layout for layout in self._db.layouts.find({}, filter)]
            return {"layouts": output}
        except ServerSelectionTimeoutError:
            return {"errors": {"server": "Could not connect to database"}}

    def add_layout(self, layout):
        """Add a layout to the database."""
        if self.get_layout(layout["name"]):
            err = f'Layout "{layout["name"]}" already exists.'
            return {"errors": {"name": err}}

        data = {"name": layout["name"], "data": layout["data"]}
        try:
            self._db.layouts.insert_one(data)
            return no_errors()
        except ServerSelectionTimeoutError:
            return {"errors": {"server": "Could not connect to database"}}

    def delete_layout(self, layout):
        """Remove a layout from the database."""
        try:
            self._db.layouts.delete_one({"name": layout})
            return no_errors()
        except ServerSelectionTimeoutError:
            return {"errors": {"server": "Could not connect to database"}}

    def _get_and_format_points(self, points, collection, filter):
        """Get data points for a specified filter and add to dict in place."""
        for item in collection.find(filter):
            time = item["time"].strftime(TIME_FORMAT_NO_TZ)
            points[time] = item["reading"]

    def get_points(self, catname, data):
        """Get data points for a specific time or time range."""
        try:
            if self.get_category(catname) is None:
                return {"errors": f'Category "{catname}" does not exist.'}

            points = {}
            collection = self._db[f"catdata_default_{catname}"]

            if "times" in data:
                data["times"] = [
                    datetime.strptime(time_str, TIME_FORMAT)
                    for time_str in data["times"]
                ]
                times = data["times"]
                filter = {"time": {"$in": times}}
                self._get_and_format_points(points, collection, filter)
            if "range" in data:
                start = datetime.strptime(data["range"]["start"], TIME_FORMAT)
                end = datetime.strptime(data["range"]["end"], TIME_FORMAT)
                filter = {"time": {"$gte": start, "$lte": end}}
                self._get_and_format_points(points, collection, filter)
            if "since" in data:
                start = datetime.strptime(data["since"], TIME_FORMAT)
                filter = {"time": {"$gte": start}}
                self._get_and_format_points(points, collection, filter)

            return {"errors": None, "points": points}
        except ServerSelectionTimeoutError:
            return {"errors": {"server": "Could not connect to database"}}

    def remove_points(self, catname, data):
        """Remove datapoints from the database."""
        try:
            if self.get_category(catname) is None:
                return {"errors": f'Category "{catname}" does not exist.'}

            output = {"removed_count": 0}
            cat_data = self._db[f"catdata_default_{catname}"]

            if "times" in data:
                times = [
                    datetime.strptime(time_str, TIME_FORMAT)
                    for time_str in data["times"]
                ]
                res = cat_data.delete_many({"time": {"$in": times}})
                output["removed_count"] += res.deleted_count
                data["times"] = [
                    (dt - dt.utcoffset()).strftime(TIME_FORMAT_NO_TZ)
                    for dt in times
                ]
            if "range" in data:
                start = datetime.strptime(data["range"]["start"], TIME_FORMAT)
                end = datetime.strptime(data["range"]["end"], TIME_FORMAT)
                res = cat_data.delete_many(
                    {"time": {"$gte": start, "$lte": end}}
                )
                output["removed_count"] += res.deleted_count
                start = (start - start.utcoffset()).strftime(TIME_FORMAT_NO_TZ)
                end = (end - end.utcoffset()).strftime(TIME_FORMAT_NO_TZ)
                data["range"]["start"] = start
                data["range"]["end"] = end
            if "since" in data:
                since = datetime.strptime(data["since"], TIME_FORMAT)
                res = cat_data.delete_many({"time": {"$gte": since}})
                output["removed_count"] += res.deleted_count
                since = (since - since.utcoffset()).strftime(TIME_FORMAT_NO_TZ)
                data["since"] = since

            output["errors"] = None
            return output
        except ServerSelectionTimeoutError:
            return {"errors": {"server": "Could not connect to database"}}

    def remove_all_points(self, catname):
        """Remove datapoints from the database."""
        cat_data = f"catdata_default_{catname}"
        try:
            res = self._db[cat_data].delete_many({})
            return {"errors": None, "removed_count": res.deleted_count}
        except ServerSelectionTimeoutError:
            return {"errors": {"server": "Could not connect to database"}}

    def get_points_range_cats(self, data):
        """Get all valid points for when a frontend category is added."""
        try:
            cat_points, now = defaultdict(dict), datetime.now(timezone.utc)
            for range_data in data:
                range = range_data["range"]
                for catname in range_data["categories"]:
                    if self.get_category(catname) is None:
                        continue

                    collection = self._db[f"catdata_default_{catname}"]
                    rtype, points = range["range_type"], {}
                    if rtype == "past":
                        amnt, unit = range["past_amount"], range["past_unit"]
                        delta = timedelta(seconds=amnt * TIME_MULTS[unit])
                        filter = {"time": {"$gte": now - delta}}
                    elif rtype == "since":
                        filter = {"time": {"$gte": range["since"]}}
                    elif rtype == "timerange":
                        start, end = range["start"], range["end"]
                        filter = {"time": {"$gte": start, "$lte": end}}

                    self._get_and_format_points(points, collection, filter)
                    cat_points[catname].update(points)

            return cat_points
        except ServerSelectionTimeoutError:
            return {"errors": {"server": "Could not connect to database"}}
